File: typescript/src/hypercore.ts
/**
 * HyperCore JSON-RPC API Client — Blocks, trading, and real-time data.
 *
 * Access HyperCore-specific methods via QuickNode's /hypercore endpoint.
 *
 * Example:
 *     import { HyperCore } from 'hyperliquid-sdk';
 *     const hc = new HyperCore("https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN");
 *     console.log(await hc.latestBlockNumber());  // Get latest block
 *     console.log(await hc.latestTrades({ count: 10 }));  // Get recent trades
 */

import { HyperliquidError } from './errors';

export interface HyperCoreOptions {
  timeout?: number;
}

/**
 * HyperCore JSON-RPC API — Blocks, trading, and real-time data.
 *
 * Access block data, trading operations, and real-time streams via JSON-RPC.
 *
 * Examples:
 *     const hc = new HyperCore("https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN");
 *
 *     // Block data
 *     await hc.latestBlockNumber();
 *     await hc.getBlock(12345);
 *     await hc.getBatchBlocks(100, 102);
 *     await hc.latestBlocks({ count: 10 });
 *
 *     // Recent data (alternative to Info methods not on QuickNode)
 *     await hc.latestTrades({ count: 10 });
 *     await hc.latestOrders({ count: 10 });
 *     await hc.latestBookUpdates({ count: 10 });
 *
 *     // Discovery
 *     await hc.listDexes();
 *     await hc.listMarkets();
 *     await hc.listMarkets({ dex: "hyperliquidity" });
 */
export class HyperCore {
  private readonly _hypercoreUrl: string;
  private readonly _timeout: number;
  private _id: number = 0;

  constructor(endpoint: string, options: HyperCoreOptions = {}) {
    this._hypercoreUrl = this._buildUrl(endpoint);
    this._timeout = options.timeout ?? 30000;
  }

  private _buildUrl(url: string): string {
    const parsed = new URL(url);
    const base = `${parsed.protocol}//${parsed.host}`;
    const pathParts = parsed.pathname.split('/').filter((p) => p.length > 0);

    // Find the token (not a known path)
    const knownPaths = new Set(['info', 'hypercore', 'evm', 'nanoreth', 'ws']);
    let token: string | null = null;
    for (const part of pathParts) {
      if (!knownPaths.has(part)) {
        token = part;
        break;
      }
    }

    if (token) {
      return `${base}/${token}/hypercore`;
    }
    return `${base}/hypercore`;
  }

  /** Make a JSON-RPC 2.0 request. */
  private async _rpc<T = unknown>(
    method: string,
    params: unknown = {}
  ): Promise<T> {
    this._id += 1;
    const body = {
      jsonrpc: '2.0',
      method,
      id: this._id,
      params,
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this._timeout);

    try {
      const response = await fetch(this._hypercoreUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new HyperliquidError(
          `Request failed with status ${response.status}`,
          {
            code: 'HTTP_ERROR',
            raw: { status: response.status, body: await response.text() },
          }
        );
      }

      const data = await response.json() as Record<string, unknown>;

      if (data.error) {
        const errorObj = data.error as Record<string, unknown>;
        const message =
          typeof errorObj === 'object' ? (errorObj.message as string) ?? String(errorObj) : String(errorObj);
        const code =
          typeof errorObj === 'object' ? String(errorObj.code ?? 'RPC_ERROR') : 'RPC_ERROR';
        throw new HyperliquidError(message, { code, raw: errorObj });
      }

      return data.result as T;
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof HyperliquidError) throw error;

      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new HyperliquidError(`Request timed out after ${this._timeout}ms`, {
            code: 'TIMEOUT',
            raw: { method, timeout: this._timeout },
          });
        }
        throw new HyperliquidError(`Connection failed: ${error.message}`, {
          code: 'CONNECTION_ERROR',
          raw: { method, error: error.message },
        });
      }
      throw error;
    }
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // BLOCK DATA
  // ═══════════════════════════════════════════════════════════════════════════

  /**
   * Get the latest block number for a stream.
   *
   * @param stream - Stream type ("trades", "orders", "events", "book", "twap", "writer_actions")
   */
  async latestBlockNumber(stream: string = 'trades'): Promise<number> {
    return this._rpc('hl_getLatestBlockNumber', { stream });
  }

  /**
   * Get a specific block by number.
   *
   * @param blockNumber - Block number to fetch
   * @param stream - Stream type ("trades", "orders", "events", "book", "twap", "writer_actions")
   */
  async getBlock(
    blockNumber: number,
    stream: string = 'trades'
  ): Promise<Record<string, unknown>> {
    // Uses array format: [stream, block_number]
    return this._rpc('hl_getBlock', [stream, blockNumber]);
  }

  /**
   * Get a range of blocks.
   *
   * @param fromBlock - Starting block number
   * @param toBlock - Ending block number (inclusive)
   * @param stream - Stream type ("trades", "orders", "events", "book", "twap", "writer_actions")
   */
  async getBatchBlocks(
    fromBlock: number,
    toBlock: number,
    stream: string = 'trades'
  ): Promise<Array<Record<string, unknown>>> {
    return this._rpc('hl_getBatchBlocks', { stream, from: fromBlock, to: toBlock });
  }

  /**
   * Get the latest blocks for a stream.
   *
   * @param stream - Stream type ("trades", "orders", "book_updates", "twap", "events", "writer_actions")
   * @param options.count - Number of blocks to return (default: 10)
   */
  async latestBlocks(
    stream: string = 'trades',
    options: { count?: number } = {}
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_getLatestBlocks', { stream, count: options.count ?? 10 });
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // RECENT DATA (Alternative to unsupported Info methods)
  // ═══════════════════════════════════════════════════════════════════════════

  /**
   * Get recent trades from latest blocks.
   *
   * This is an alternative to Info.recentTrades() which is not available on QuickNode.
   *
   * @param options.count - Number of blocks to fetch (default: 10)
   * @param options.coin - Optional coin filter (e.g., "BTC", "ETH")
   */
  async latestTrades(
    options: { count?: number; coin?: string } = {}
  ): Promise<Array<Record<string, unknown>>> {
    const result = (await this.latestBlocks('trades', {
      count: options.count ?? 10,
    })) as { blocks?: Array<{ events?: Array<unknown[]> }> };
    const trades: Array<Record<string, unknown>> = [];

    for (const block of result.blocks ?? []) {
      for (const event of block.events ?? []) {
        if (Array.isArray(event) && event.length >= 2) {
          const user = event[0] as string;
          const trade = event[1] as Record<string, unknown>;
          if (options.coin === undefined || trade.coin === options.coin) {
            trades.push({ user, ...trade });
          }
        }
      }
    }
    return trades;
  }

  /**
   * Get recent order events from latest blocks.
   *
   * @param options.count - Number of blocks to fetch (default: 10)
   * @param options.coin - Optional coin filter
   */
  async latestOrders(
    options: { count?: number; coin?: string } = {}
  ): Promise<Array<Record<string, unknown>>> {
    const result = (await this.latestBlocks('orders', {
      count: options.count ?? 10,
    })) as { blocks?: Array<{ events?: Array<unknown[]> }> };
    const orders: Array<Record<string, unknown>> = [];

    for (const block of result.blocks ?? []) {
      for (const event of block.events ?? []) {
        if (Array.isArray(event) && event.length >= 2) {
          const user = event[0] as string;
          const order = event[1] as Record<string, unknown>;
          if (options.coin === undefined || order.coin === options.coin) {
            orders.push({ user, ...order });
          }
        }
      }
    }
    return orders;
  }

  /**
   * Get recent book updates from latest blocks.
   *
   * This is an alternative to Info.l2Book() for real-time updates.
   *
   * @param options.count - Number of blocks to fetch (default: 10)
   * @param options.coin - Optional coin filter
   */
  async latestBookUpdates(
    options: { count?: number; coin?: string } = {}
  ): Promise<Array<Record<string, unknown>>> {
    const result = (await this.latestBlocks('book_updates', {
      count: options.count ?? 10,
    })) as { blocks?: Array<{ events?: Array<Record<string, unknown>> }> };
    const updates: Array<Record<string, unknown>> = [];

    for (const block of result.blocks ?? []) {
      for (const event of block.events ?? []) {
        if (typeof event === 'object' && event !== null) {
          if (options.coin === undefined || event.coin === options.coin) {
            updates.push(event);
          }
        }
      }
    }
    return updates;
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // DISCOVERY
  // ═══════════════════════════════════════════════════════════════════════════

  /** List all available DEXes. */
  async listDexes(): Promise<Array<Record<string, unknown>>> {
    return this._rpc('hl_listDexes');
  }

  /**
   * List available markets.
   *
   * @param options.dex - Optional DEX filter (e.g., "hyperliquidity")
   */
  async listMarkets(
    options: { dex?: string } = {}
  ): Promise<Array<Record<string, unknown>>> {
    const params = options.dex ? { dex: options.dex } : undefined;
    return this._rpc('hl_listMarkets', params);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // ORDER QUERIES
  // ═══════════════════════════════════════════════════════════════════════════

  /** Get open orders for a user. */
  async openOrders(user: string): Promise<Array<Record<string, unknown>>> {
    return this._rpc('hl_openOrders', { user });
  }

  /** Get status of a specific order. */
  async orderStatus(user: string, oid: number): Promise<Record<string, unknown>> {
    return this._rpc('hl_orderStatus', { user, oid });
  }

  /**
   * Validate an order before signing (preflight check).
   *
   * @param coin - Trading pair (e.g., "BTC")
   * @param isBuy - True for buy, False for sell
   * @param limitPx - Limit price as string
   * @param sz - Size as string
   * @param user - User address
   * @param options.reduceOnly - Whether order is reduce-only
   * @param options.orderType - Optional order type configuration
   */
  async preflight(
    coin: string,
    isBuy: boolean,
    limitPx: string,
    sz: string,
    user: string,
    options: { reduceOnly?: boolean; orderType?: Record<string, unknown> } = {}
  ): Promise<Record<string, unknown>> {
    const params: Record<string, unknown> = {
      coin,
      isBuy,
      limitPx,
      sz,
      user,
      reduceOnly: options.reduceOnly ?? false,
    };
    if (options.orderType) {
      params.orderType = options.orderType;
    }
    return this._rpc('hl_preflight', params);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // BUILDER FEE
  // ═══════════════════════════════════════════════════════════════════════════

  /** Get maximum builder fee for a user-builder pair. */
  async getMaxBuilderFee(
    user: string,
    builder: string
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_getMaxBuilderFee', { user, builder });
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // ORDER BUILDING (Returns unsigned actions for signing)
  // ═══════════════════════════════════════════════════════════════════════════

  /**
   * Build an order for signing.
   *
   * @param coin - Trading pair (e.g., "BTC")
   * @param isBuy - True for buy, False for sell
   * @param limitPx - Limit price as string
   * @param sz - Size as string
   * @param user - User address
   * @param options.reduceOnly - Whether order is reduce-only
   * @param options.orderType - Optional order type configuration
   * @param options.cloid - Optional client order ID
   */
  async buildOrder(
    coin: string,
    isBuy: boolean,
    limitPx: string,
    sz: string,
    user: string,
    options: {
      reduceOnly?: boolean;
      orderType?: Record<string, unknown>;
      cloid?: string;
    } = {}
  ): Promise<Record<string, unknown>> {
    const params: Record<string, unknown> = {
      coin,
      isBuy,
      limitPx,
      sz,
      user,
      reduceOnly: options.reduceOnly ?? false,
    };
    if (options.orderType) {
      params.orderType = options.orderType;
    }
    if (options.cloid) {
      params.cloid = options.cloid;
    }
    return this._rpc('hl_buildOrder', params);
  }

  /**
   * Build a cancel action for signing.
   *
   * @param coin - Trading pair
   * @param oid - Order ID to cancel
   * @param user - User address
   */
  async buildCancel(
    coin: string,
    oid: number,
    user: string
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_buildCancel', { coin, oid, user });
  }

  /**
   * Build a modify action for signing.
   *
   * @param coin - Trading pair
   * @param oid - Order ID to modify
   * @param user - User address
   * @param options.limitPx - New limit price
   * @param options.sz - New size
   * @param options.isBuy - New side
   */
  async buildModify(
    coin: string,
    oid: number,
    user: string,
    options: { limitPx?: string; sz?: string; isBuy?: boolean } = {}
  ): Promise<Record<string, unknown>> {
    const params: Record<string, unknown> = { coin, oid, user };
    if (options.limitPx !== undefined) {
      params.limitPx = options.limitPx;
    }
    if (options.sz !== undefined) {
      params.sz = options.sz;
    }
    if (options.isBuy !== undefined) {
      params.isBuy = options.isBuy;
    }
    return this._rpc('hl_buildModify', params);
  }

  /**
   * Build a builder fee approval for signing.
   *
   * @param user - User address
   * @param builder - Builder address
   * @param maxFeeRate - Maximum fee rate (e.g., "0.001" for 0.1%)
   * @param nonce - Nonce for the action
   */
  async buildApproveBuilderFee(
    user: string,
    builder: string,
    maxFeeRate: string,
    nonce: number
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_buildApproveBuilderFee', {
      user,
      builder,
      maxFeeRate,
      nonce,
    });
  }

  /**
   * Build a builder fee revocation for signing.
   *
   * @param user - User address
   * @param builder - Builder address to revoke
   * @param nonce - Nonce for the action
   */
  async buildRevokeBuilderFee(
    user: string,
    builder: string,
    nonce: number
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_buildRevokeBuilderFee', { user, builder, nonce });
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // SENDING SIGNED ACTIONS
  // ═══════════════════════════════════════════════════════════════════════════

  /**
   * Send a signed order.
   *
   * @param action - The order action from buildOrder
   * @param signature - EIP-712 signature
   * @param nonce - Nonce used in signing
   */
  async sendOrder(
    action: Record<string, unknown>,
    signature: string,
    nonce: number
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_sendOrder', { action, signature, nonce });
  }

  /**
   * Send a signed cancel.
   *
   * @param action - The cancel action from buildCancel
   * @param signature - EIP-712 signature
   * @param nonce - Nonce used in signing
   */
  async sendCancel(
    action: Record<string, unknown>,
    signature: string,
    nonce: number
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_sendCancel', { action, signature, nonce });
  }

  /**
   * Send a signed modify.
   *
   * @param action - The modify action from buildModify
   * @param signature - EIP-712 signature
   * @param nonce - Nonce used in signing
   */
  async sendModify(
    action: Record<string, unknown>,
    signature: string,
    nonce: number
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_sendModify', { action, signature, nonce });
  }

  /**
   * Send a signed builder fee approval.
   *
   * @param action - The approval action from buildApproveBuilderFee
   * @param signature - EIP-712 signature
   */
  async sendApproval(
    action: Record<string, unknown>,
    signature: string
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_sendApproval', { action, signature });
  }

  /**
   * Send a signed builder fee revocation.
   *
   * @param action - The revocation action from buildRevokeBuilderFee
   * @param signature - EIP-712 signature
   */
  async sendRevocation(
    action: Record<string, unknown>,
    signature: string
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_sendRevocation', { action, signature });
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // WEBSOCKET SUBSCRIPTIONS (via JSON-RPC)
  // ═══════════════════════════════════════════════════════════════════════════

  /**
   * Subscribe to a WebSocket stream via JSON-RPC.
   *
   * @param subscription - Subscription parameters (type, coin, user, etc.)
   *
   * Example:
   *     await hc.subscribe({ type: "trades", coin: "BTC" });
   */
  async subscribe(
    subscription: Record<string, unknown>
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_subscribe', { subscription });
  }

  /**
   * Unsubscribe from a WebSocket stream.
   *
   * @param subscription - Subscription parameters to unsubscribe from
   */
  async unsubscribe(
    subscription: Record<string, unknown>
  ): Promise<Record<string, unknown>> {
    return this._rpc('hl_unsubscribe', { subscription });
  }
}

File: rust/src/hypercore.rs
//! HyperCore API client for Hyperliquid.
//!
//! Provides access to block-level data, trades, orders, and book updates.

use serde_json::{json, Value};
use std::sync::Arc;

use crate::client::HyperliquidSDKInner;
use crate::error::Result;

/// HyperCore API client
pub struct HyperCore {
    inner: Arc<HyperliquidSDKInner>,
}

impl HyperCore {
    pub(crate) fn new(inner: Arc<HyperliquidSDKInner>) -> Self {
        Self { inner }
    }

    /// Get the HyperCore endpoint URL
    fn hypercore_url(&self) -> String {
        self.inner.hypercore_url()
    }

    /// Make a JSON-RPC request to HyperCore
    async fn rpc(&self, method: &str, params: Value) -> Result<Value> {
        let url = self.hypercore_url();

        let body = json!({
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
            "id": 1,
        });

        let response = self
            .inner
            .http_client
            .post(&url)
            .json(&body)
            .send()
            .await?;

        let status = response.status();
        let text = response.text().await?;

        if !status.is_success() {
            return Err(crate::Error::NetworkError(format!(
                "HyperCore request failed {}: {}",
                status, text
            )));
        }

        let result: Value = serde_json::from_str(&text)?;

        // Check for JSON-RPC error
        if let Some(error) = result.get("error") {
            let message = error
                .get("message")
                .and_then(|m| m.as_str())
                .unwrap_or("Unknown error");
            return Err(crate::Error::ApiError {
                code: crate::error::ErrorCode::Unknown,
                message: message.to_string(),
                guidance: "Check your HyperCore request parameters.".to_string(),
                raw: Some(error.to_string()),
            });
        }

        Ok(result.get("result").cloned().unwrap_or(Value::Null))
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Block Data
    // ──────────────────────────────────────────────────────────────────────────

    /// Get the latest block number for a stream
    pub async fn latest_block_number(&self, stream: Option<&str>) -> Result<u64> {
        let stream = stream.unwrap_or("trades");
        let result = self.rpc("hl_getLatestBlockNumber", json!({"stream": stream})).await?;
        result
            .as_u64()
            .ok_or_else(|| crate::Error::JsonError("Invalid block number".to_string()))
    }

    /// Get a specific block
    pub async fn get_block(&self, block_number: u64, stream: Option<&str>) -> Result<Value> {
        let stream = stream.unwrap_or("trades");
        self.rpc("hl_getBlock", json!([stream, block_number])).await
    }

    /// Get a batch of blocks
    pub async fn get_batch_blocks(
        &self,
        from_block: u64,
        to_block: u64,
        stream: Option<&str>,
    ) -> Result<Value> {
        let stream = stream.unwrap_or("trades");
        self.rpc("hl_getBatchBlocks", json!({"stream": stream, "from": from_block, "to": to_block}))
            .await
    }

    /// Get latest blocks
    pub async fn latest_blocks(&self, stream: Option<&str>, count: Option<u32>) -> Result<Value> {
        let stream = stream.unwrap_or("trades");
        let count = count.unwrap_or(10);
        self.rpc("hl_getLatestBlocks", json!({"stream": stream, "count": count})).await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Recent Data
    // ──────────────────────────────────────────────────────────────────────────

    /// Get latest trades from recent blocks
    /// This is an alternative to Info.recent_trades() for QuickNode endpoints
    pub async fn latest_trades(&self, count: Option<u32>, coin: Option<&str>) -> Result<Value> {
        let count = count.unwrap_or(10);
        let blocks = self.latest_blocks(Some("trades"), Some(count)).await?;

        let mut trades = Vec::new();
        if let Some(blocks_arr) = blocks.get("blocks").and_then(|b| b.as_array()) {
            for block in blocks_arr {
                if let Some(events) = block.get("events").and_then(|e| e.as_array()) {
                    for event in events {
                        if let Some(arr) = event.as_array() {
                            if arr.len() >= 2 {
                                let user = &arr[0];
                                if let Some(trade) = arr.get(1) {
                                    // Apply coin filter if specified
                                    if let Some(c) = coin {
                                        if trade.get("coin").and_then(|tc| tc.as_str()) != Some(c) {
                                            continue;
                                        }
                                    }
                                    let mut trade_obj = trade.clone();
                                    if let Some(obj) = trade_obj.as_object_mut() {
                                        obj.insert("user".to_string(), user.clone());
                                    }
                                    trades.push(trade_obj);
                                }
                            }
                        }
                    }
                }
            }
        }
        Ok(json!(trades))
    }

    /// Get latest orders from recent blocks
    pub async fn latest_orders(&self, count: Option<u32>) -> Result<Value> {
        let count = count.unwrap_or(10);
        let blocks = self.latest_blocks(Some("orders"), Some(count)).await?;

        let mut orders = Vec::new();
        if let Some(blocks_arr) = blocks.get("blocks").and_then(|b| b.as_array()) {
            for block in blocks_arr {
                if let Some(events) = block.get("events").and_then(|e| e.as_array()) {
                    for event in events {
                        if let Some(arr) = event.as_array() {
                            if arr.len() >= 2 {
                                let user = &arr[0];
                                if let Some(order) = arr.get(1) {
                                    let mut order_obj = order.clone();
                                    if let Some(obj) = order_obj.as_object_mut() {
                                        obj.insert("user".to_string(), user.clone());
                                    }
                                    orders.push(order_obj);
                                }
                            }
                        }
                    }
                }
            }
        }
        Ok(json!(orders))
    }

    /// Get latest book updates from recent blocks
    pub async fn latest_book_updates(&self, count: Option<u32>, coin: Option<&str>) -> Result<Value> {
        let count = count.unwrap_or(10);
        let blocks = self.latest_blocks(Some("book_updates"), Some(count)).await?;

        let mut updates = Vec::new();
        if let Some(blocks_arr) = blocks.get("blocks").and_then(|b| b.as_array()) {
            for block in blocks_arr {
                if let Some(events) = block.get("events").and_then(|e| e.as_array()) {
                    for event in events {
                        // Apply coin filter if specified
                        if let Some(c) = coin {
                            if event.get("coin").and_then(|ec| ec.as_str()) != Some(c) {
                                continue;
                            }
                        }
                        updates.push(event.clone());
                    }
                }
            }
        }
        Ok(json!(updates))
    }

    /// Get latest TWAP updates
    pub async fn latest_twap(&self, count: Option<u32>) -> Result<Value> {
        let count = count.unwrap_or(10);
        self.latest_blocks(Some("twap"), Some(count)).await
    }

    /// Get latest events
    pub async fn latest_events(&self, count: Option<u32>) -> Result<Value> {
        let count = count.unwrap_or(10);
        self.latest_blocks(Some("events"), Some(count)).await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Discovery
    // ──────────────────────────────────────────────────────────────────────────

    /// List all DEXes
    pub async fn list_dexes(&self) -> Result<Value> {
        self.rpc("hl_listDexes", json!({})).await
    }

    /// List all markets (optionally for a specific DEX)
    pub async fn list_markets(&self, dex: Option<&str>) -> Result<Value> {
        if let Some(d) = dex {
            self.rpc("hl_listMarkets", json!({"dex": d})).await
        } else {
            self.rpc("hl_listMarkets", json!({})).await
        }
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Order Queries
    // ──────────────────────────────────────────────────────────────────────────

    /// Get open orders for a user
    pub async fn open_orders(&self, user: &str) -> Result<Value> {
        self.rpc("hl_openOrders", json!({"user": user})).await
    }

    /// Get status of a specific order
    pub async fn order_status(&self, user: &str, oid: u64) -> Result<Value> {
        self.rpc("hl_orderStatus", json!({"user": user, "oid": oid}))
            .await
    }

    /// Validate an order before signing (preflight check)
    pub async fn preflight(
        &self,
        coin: &str,
        is_buy: bool,
        limit_px: &str,
        sz: &str,
        user: &str,
        reduce_only: bool,
        order_type: Option<&Value>,
    ) -> Result<Value> {
        let mut params = json!({
            "coin": coin,
            "isBuy": is_buy,
            "limitPx": limit_px,
            "sz": sz,
            "user": user,
            "reduceOnly": reduce_only,
        });
        if let Some(ot) = order_type {
            params["orderType"] = ot.clone();
        }
        self.rpc("hl_preflight", params).await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Builder Fee
    // ──────────────────────────────────────────────────────────────────────────

    /// Get maximum builder fee for a user-builder pair
    pub async fn get_max_builder_fee(&self, user: &str, builder: &str) -> Result<Value> {
        self.rpc("hl_getMaxBuilderFee", json!({"user": user, "builder": builder}))
            .await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Order Building (Returns unsigned actions for signing)
    // ──────────────────────────────────────────────────────────────────────────

    /// Build an order for signing
    pub async fn build_order(
        &self,
        coin: &str,
        is_buy: bool,
        limit_px: &str,
        sz: &str,
        user: &str,
        reduce_only: bool,
        order_type: Option<&Value>,
        cloid: Option<&str>,
    ) -> Result<Value> {
        let mut params = json!({
            "coin": coin,
            "isBuy": is_buy,
            "limitPx": limit_px,
            "sz": sz,
            "user": user,
            "reduceOnly": reduce_only,
        });
        if let Some(ot) = order_type {
            params["orderType"] = ot.clone();
        }
        if let Some(c) = cloid {
            params["cloid"] = json!(c);
        }
        self.rpc("hl_buildOrder", params).await
    }

    /// Build a cancel action for signing
    pub async fn build_cancel(&self, coin: &str, oid: u64, user: &str) -> Result<Value> {
        self.rpc("hl_buildCancel", json!({"coin": coin, "oid": oid, "user": user}))
            .await
    }

    /// Build a modify action for signing
    pub async fn build_modify(
        &self,
        coin: &str,
        oid: u64,
        user: &str,
        limit_px: Option<&str>,
        sz: Option<&str>,
        is_buy: Option<bool>,
    ) -> Result<Value> {
        let mut params = json!({"coin": coin, "oid": oid, "user": user});
        if let Some(px) = limit_px {
            params["limitPx"] = json!(px);
        }
        if let Some(s) = sz {
            params["sz"] = json!(s);
        }
        if let Some(buy) = is_buy {
            params["isBuy"] = json!(buy);
        }
        self.rpc("hl_buildModify", params).await
    }

    /// Build a builder fee approval for signing
    pub async fn build_approve_builder_fee(
        &self,
        user: &str,
        builder: &str,
        max_fee_rate: &str,
        nonce: u64,
    ) -> Result<Value> {
        self.rpc(
            "hl_buildApproveBuilderFee",
            json!({
                "user": user,
                "builder": builder,
                "maxFeeRate": max_fee_rate,
                "nonce": nonce,
            }),
        )
        .await
    }

    /// Build a builder fee revocation for signing
    pub async fn build_revoke_builder_fee(
        &self,
        user: &str,
        builder: &str,
        nonce: u64,
    ) -> Result<Value> {
        self.rpc(
            "hl_buildRevokeBuilderFee",
            json!({
                "user": user,
                "builder": builder,
                "nonce": nonce,
            }),
        )
        .await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Sending Signed Actions
    // ──────────────────────────────────────────────────────────────────────────

    /// Send a signed order
    pub async fn send_order(
        &self,
        action: &Value,
        signature: &str,
        nonce: u64,
    ) -> Result<Value> {
        self.rpc(
            "hl_sendOrder",
            json!({"action": action, "signature": signature, "nonce": nonce}),
        )
        .await
    }

    /// Send a signed cancel
    pub async fn send_cancel(
        &self,
        action: &Value,
        signature: &str,
        nonce: u64,
    ) -> Result<Value> {
        self.rpc(
            "hl_sendCancel",
            json!({"action": action, "signature": signature, "nonce": nonce}),
        )
        .await
    }

    /// Send a signed modify
    pub async fn send_modify(
        &self,
        action: &Value,
        signature: &str,
        nonce: u64,
    ) -> Result<Value> {
        self.rpc(
            "hl_sendModify",
            json!({"action": action, "signature": signature, "nonce": nonce}),
        )
        .await
    }

    /// Send a signed builder fee approval
    pub async fn send_approval(&self, action: &Value, signature: &str) -> Result<Value> {
        self.rpc(
            "hl_sendApproval",
            json!({"action": action, "signature": signature}),
        )
        .await
    }

    /// Send a signed builder fee revocation
    pub async fn send_revocation(&self, action: &Value, signature: &str) -> Result<Value> {
        self.rpc(
            "hl_sendRevocation",
            json!({"action": action, "signature": signature}),
        )
        .await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // WebSocket Subscriptions (via JSON-RPC)
    // ──────────────────────────────────────────────────────────────────────────

    /// Subscribe to a WebSocket stream via JSON-RPC
    pub async fn subscribe(&self, subscription: &Value) -> Result<Value> {
        self.rpc("hl_subscribe", json!({"subscription": subscription}))
            .await
    }

    /// Unsubscribe from a WebSocket stream
    pub async fn unsubscribe(&self, subscription: &Value) -> Result<Value> {
        self.rpc("hl_unsubscribe", json!({"subscription": subscription}))
            .await
    }
}
